src/mmw/apps/modeling/mapshed/tasks.py

```diff
@@ -244,41 +247,45 @@ def nlcd_streams_drb(result):
 
 
 @shared_task(throws=Exception)
-def nlcd_soils(result):
+def nlcd_soil(result):
     """
     Results are expected to be in the format:
     {
-      (NLCD ID, Soil Group ID, Soil Texture ID): Count,
+      (NLCD ID, Soil Group ID): Count,
     }
 
     We calculate a number of values relying on various combinations
     of these raster datasets.
     """
     if 'error' in result:
-        raise Exception('[nlcd_soils] {}'.format(result['error']))
+        raise Exception('[nlcd_soil] {}'.format(result['error']))
 
-    ngt_count = parse(result)
+    ng_count = parse(result)
 
     # Raise exception if no NLCD values
-    if len(ngt_count.values()) == 0:
+    if len(ng_count.values()) == 0:
         raise Exception(NO_LAND_COVER)
 
-    # Split combined counts into separate ones for processing
-    # Reduce [(n, g, t): c] to
-    n_count = {}   # [n: sum(c)]
-    ng_count = {}  # [(n, g): sum(c)]
-    nt_count = {}  # [(n, t): sum(c)]
-    for (n, g, t), count in ngt_count.iteritems():
-        n_count[n] = count + n_count.get(n, 0)
-        nt_count[(n, t)] = count + nt_count.get((n, t), 0)
+    # Get a list of the counts for a list of ng_count keys
+    def counts(keys):
+        return [ng_count[key] for key in keys]
+
+    # Split combined counts into separate count dict for processing
+    # Takes a lambda of the form lambda(n,g): <what you want to group on>
+    def count_by_group(key_func):
+        group_iter = groupby(sorted(ng_count, key=key_func),
+                             key_func)
+        grouped = {key: list(ng_keys) for key, ng_keys in group_iter}
+        return {key: sum(counts(ng_keys))
+                for key, ng_keys in grouped.iteritems()}
 
-        # Map soil group values to usable subset
-        g2 = settings.SOIL_GROUP[g]
-        ng_count[(n, g2)] = count + ng_count.get((n, g2), 0)
+    # Group by land
+    n_count = count_by_group(lambda(n, g): n)
+    # Group by (land, usable soil group subset)
+    ng2_count = count_by_group(lambda(n, g): (n, settings.SOIL_GROUP[g]))
```

Comment:
I thought I was being clever, but this is actually slower than the for loop. About to push a fixup to put it back.